codex/blockexchange/engine/discovery.nim

```diff
@@ -152,14 +155,22 @@ proc discoveryTaskLoop(b: DiscoveryEngine) {.async.} =
           let
             peers = await request
 
-          trace "Discovered peers", peers = peers.len
+          trace "Discovered peers for block", peers = peers.len, cid
           let
             dialed = await allFinished(
               peers.mapIt( b.network.dialPeer(it.data) ))
 
           for i, f in dialed:
             if f.failed:
+              trace "dial failed"
               await b.discovery.removeProvider(peers[i].data.peerId)
+            else:
+              trace "dial successful", peerId = peers[i].data.peerId
+
+              without cb =? b.blockLocatedCallback:
+                trace "blockLocated callback not set :o"
+
+              await cb(cid, peers[i].data.peerId)
```

Comment:
I'm not sure you need this callback, there are several reasons why:

- Doing anything with the peer only makes sense after the peer has been fully connected, which only happens after the `PeerEventKind.Joined` has been triggered
- The logic in the callback is already triggered in `setupPeer`, which is registered for the `PeerEventKind.Joined` event. For example the local block list for the peer's "Haves" should be updated on peer connect, which happens in `setupPeer` again

In other words, simply dialing the peer should be enough - it either succeeds upgrading and is fully connected, or it fails in which case there isn't much to be done.